### PR TITLE
Work towards introducing byte values

### PR DESCRIPTION
**Pull request by tomhoule**

### CODE REVIEW

File: src/ast/values.rs [hunk 1]
```diff
@@ -99,6 +101,7 @@ impl<'a> From<ParameterizedValue<'a>> for Value {
             ParameterizedValue::Integer(i) => Value::Number(Number::from(i)),
             ParameterizedValue::Real(d) => serde_json::to_value(d).unwrap(),
             ParameterizedValue::Text(cow) => Value::String(cow.into_owned()),
+            ParameterizedValue::Bytes(bytes) => todo!("bytes input"),
```
Comment (tomhoule) [inline src/ast/values.rs]: This needs to be addressed.

File: src/ast/values.rs [hunk 2]
```diff
@@ -64,6 +65,7 @@ impl<'a> fmt::Display for ParameterizedValue<'a> {
             ParameterizedValue::Integer(val) => write!(f, "{}", val),
             ParameterizedValue::Real(val) => write!(f, "{}", val),
             ParameterizedValue::Text(val) => write!(f, "\"{}\"", val),
+            ParameterizedValue::Bytes(_bytes) => write!(f, "<bytes>"),
```
Comment (tomhoule) [inline src/ast/values.rs]: This needs to be addressed.

File: src/ast/values.rs [hunk 3]
```diff
@@ -99,6 +101,7 @@ impl<'a> From<ParameterizedValue<'a>> for Value {
             ParameterizedValue::Integer(i) => Value::Number(Number::from(i)),
             ParameterizedValue::Real(d) => serde_json::to_value(d).unwrap(),
             ParameterizedValue::Text(cow) => Value::String(cow.into_owned()),
+            ParameterizedValue::Bytes(bytes) => Value::String(String::from_utf8_lossy(&bytes).into_owned()),
```
Comment (pimeys) [inline src/ast/values.rs]: hmm

File: src/ast/values.rs [hunk 4]
```diff
@@ -165,6 +170,33 @@ impl<'a> ParameterizedValue<'a> {
...
+    pub fn as_bytes(&self) -> Option<&[u8]> {
+        match self {
+            ParameterizedValue::Text(cow) => Some(cow.as_ref().as_bytes()),
+            ParameterizedValue::Bytes(cow) => Some(cow.as_ref()),
+            _ => None,
+        }
+    }
+
+    /// Returns a cloned `Vec<u8>` if the value is text or a byte slice, otherwise `None`.
+    pub fn to_bytes(&self) -> Option<Vec<u8>> {
```
Comment (pimeys) [inline src/ast/values.rs]: `into_bytes`

File: src/ast/values.rs [hunk 5]
```diff
@@ -377,6 +409,12 @@ impl<'a> From<i32> for ParameterizedValue<'a> {
     }
 }
 
+impl<'a> From<&'a [u8]> for ParameterizedValue<'a> {
```
Comment (pimeys) [inline src/ast/values.rs]: Maybe also `From<Vec<u8>>`?

File: src/connector/mysql.rs [hunk 6]
```diff
@@ -384,10 +384,42 @@ VALUES (1, 'Joe', 27, 20000.00 );
...
+
+        connection
+            .query_raw(
+                "CREATE TABLE mysql_blobs_roundtrip_test (id int AUTO_INCREMENT PRIMARY KEY, bytes MEDIUMBLOB)",
+                &[],
+            )
+            .await
+            .unwrap();
+
+        let insert = crate::ast::Insert::single_into("mysql_blobs_roundtrip_test").value("bytes", blob.as_slice());
```
Comment (pimeys) [inline src/connector/mysql.rs]: `use`!

File: src/connector/mysql.rs [hunk 7]
```diff
@@ -384,10 +384,42 @@ VALUES (1, 'Joe', 27, 20000.00 );
...
+
+        connection.query(insert.into()).await.unwrap();
+
+        let roundtripped = crate::ast::Select::from_table("mysql_blobs_roundtrip_test").column("bytes");
```
Comment (pimeys) [inline src/connector/mysql.rs]: `use`!

File: src/ast/values.rs (hunk 4 above)
Comment (tomhoule) [reply src/ast/values.rs]: by convention `into_bytes` would take ownership though, right? I think there's also an `into_bytes` already. If not I'll add it.

File: src/ast/values.rs (hunk 5 above)
Comment (tomhoule) [reply src/ast/values.rs]: It's not possible because of the `From<Vec<T>> for DatabaseValue` I think

File: src/ast/values.rs (hunk 4 above)
Comment (pimeys) [reply src/ast/values.rs]: Achso. Yes.

**Pull request merged**